### ISSUE TITLE
How to Deal with Second Edition Cards

### ISSUE THREAD

**Issue by nickv2002** [sumpfork/dominiontabs#73]
How do you want to deal with the [second edition](https://boardgamegeek.com/article/23765096#23765096) which are already shipping to stores?

Some Options I thought of:
1. New "Second Edition" sets for Dominion and Intrigue
2. New "Second Edition" Group of the new cards

**Comment by nickv2002** [comment]
Additionally, it sounds like we're getting new rules to be updated and new wording on a few cards (adding "may") that might actually change the rules for them slightly.

**Comment by sumpfork** [comment]
I think '2nd ed.' versions of Dominion and Intrigue will probably work ok. These can also have the wording changes. There'll be a good amount of data duplication, obviously - I guess we could get fancy and have sets inherit from others with overriding, adding and deleting, but I'm not sure that's worth the effort and would make it harder for people to contribute that aren't familiar with how those mechanisms work. 

We already have an example of (sub-)set duplication between the 'Dominion' and 'Base' sets.

I'm not sure we'll be able to track individual slight card text/rule updates in new printings of other sets very well, and I suspect it probably doesn't matter that much - people do have the actual rules and cards in case there are questions.

**Comment by nickv2002** [comment]
@stumpfork What about for the folks that just purchase the upgrade packs with the 7 new cards (14 total between the 2 sets)? Can we also put in an _Upgrade_ set for those?

**Comment by sumpfork** [comment]
To just print the new dividers instead of printing a whole set again, you mean? Hmm, yeah, I guess that'd be nice - in fact, I might need that :).

Well, is it worth having some basic inheritance in that case? Like a `set_definitions.json` file that lists something like
```
[
...
{
    "set": "Dominion 2nd. Edition",
    "inherits_from": "Dominion",
    "cards_to_delete": ["cardname1", "cardname2"...]
},
...
]
```
with the default being that cards from parent sets are included if not deleted, and overridden if specified again in the actual `cards.json` file for the inheriting set?

**Comment by nickv2002** [comment]
I'm fine with inheritance. But there's not a _lot_ of crossover so I'm also fine with the dumb option of just having duplicated data for cards that are in multiple sets.

(I figure many people with existing prints of the dividers will be buying the upgrade packs with the 14 new cards, myself included.)

**Comment by nickv2002** [comment]
Missed this last week, but here's all the info on the new cards as well as the removed ones for [Dominion](https://boardgamegeek.com/thread/1644466/rulebook-new-cards) ([Rulebook PDF](http://riograndegames.com/getFile.php?id=2011)) and [Intrigue](https://boardgamegeek.com/thread/1644473/rulebook-new-cards) ([Rulebook PDF](http://riograndegames.com/getFile.php?id=2012)).

**Comment by nickv2002** [comment]
Here's an updated [group of icons for every set](https://boardgamegeek.com/filepage/73273/dominion-card-icons-vector-images) including the new Second Editions.

**Comment by wvoigt** [comment]
I know this probably will not be popular, but all the card data duplication is starting to make it hard to keep everything up to date.   Just about every bit of card data is duplicated in each language. And since there is no unique key/identifier on each card that spans languages, it is hard to match cards across the languages.

I know it would be a big change, but I think the following would make things easier to maintain in the long run:

- Each card had a unique identifier/name that is used as a 'key'.
- Likewise sets would have a unique identifier.
- Data that is common across all languages could be separated out and put in one place, or stored in only one primary language file.
- Using the card identifiers, language files would only need have the text elements unique for that language.  These are generally: name, description, and extra.
- As a backup, the primary language can be used if a card or set has not been translated.

This has several advantages:
If there is any structural changes/fixes, it goes to all languages immediately.
Lets say a new card is added, then all languages at least get the divider, even if it temporarily has the primary language.  It will also make it more obvious that a translation is needed.  As it is now, it is just not there.  And until someone translates it, and adds its data, it does not even show up.

**Comment by nickv2002** [comment]
@wvoigt I think you're generally right and your changes would make things easier to maintain.  However, we have to balance all that with the work it would require to make the changes vs the cost of continuing as is. If we knew with certainty that there would be no more expansions, only new promos, would this be worth doing? What about 1 expansion? 2, 3? 

So sure, it does make sense to refactor at some point but the calculus of when to do that and who will do the work is difficult.

**Comment by sumpfork** [comment]
I do agree in principle, and note that not only future sets but also translations would greatly benefit from the increased sharing and thus error checking and consistency that the English version has received. Plus, it would be easier to error check what is missing in translations.

I also think we're not that far from doing this. Card names are mostly already a unique identifier, with only revisions requiring added tagging. Sets are already unique. That said, I do agree that tags for both should be separate from printed name strings. Separating out the common info just requires making a common JSON file with only that info and the card tag, and leaving the card name/set name/description/extra in the language-specific files. One question is whats should happen to `mapping.json`. We still need a translation of set names to readable names, but right now that file also contains the promo card title translations. I think it should be just `sets.json` and contain a mapping from set tag to set name in the specific language - promo card names should be specified in the `cards.json` file and their tags should be used to specify images.

So, a common `card_spec.json` entry should be
```
 {
  "card_tag": <card_tag>,
  "cardset_tag": <cardset_tag>,
  "cost": "4"
  "potcost": 0,
  "types":[
   "Action"
  ]
 }
```

and `cards.json` should just have
```
 {
  "card_tag": <card_tag>,
  "description": "+1 Card\n+1 Action\nYou may play an Avanto from your hand.\nWhile this is in play, when you play a Silver, you may trash a card from your hand.",
  "extra": "Sauna is a promotional Action card. It's a cantrip that allows you to trash a card when you play a Silver. It is a split pile card, with five copies of Sauna sitting on top of five copies of Avanto.",
  "name": "Sauna"
 }
```
in the appropriate language.

**Comment by nickv2002** [comment]
@sumpfork Would your proposed `card_spec.json` and `cards.json` files have any special logic or features to handle split-piles etc that have different cards under one divider?  Or would it just work the same way it does now with a separate split and combined entries?

**Comment by ghostsquad** [comment]
👍

**Comment by sumpfork** [comment]
I hadn't really thought about reworking split piles in this context, but we certainly can consider it if there's a good proposal.

**Comment by nickv2002** [comment]
@mr-ice was kind enough to type up the 2nd edition cards: https://github.com/sumpfork/dominiontabs/pull/83/commits
Still need to add the 2nd edition Intrigue Cards too.

**Comment by nickv2002** [comment]
As for split piles: let's just put every individual card in `cards.json` just like you listed including both cards from a split pile.  Then, for split pile cards we can have a special tag in the `card_spec.json` like so

 ```
{
  "card_tag": <card_tag>,
  "cardset_tag": <cardset_tag>,
  "splitpile_tag": <splitpile_tag>,
  "cost": "4"
  "potcost": 0,
  "types":[
   "Action"
  ]
 }
```

Any cards that went together in one pile would get the same `splitpile_tag`, then, when generating, if the user selected the combined option, the generator would look for anything with the same `splitpile_tag` and combine it into one entry programmatically by selecting the lowest cost and union-ing the types. This way we can get rid of extraneous entries for the combined piles. It should work for stuff like the _Cornucopia Prizes_ too. Does this sound reasonable?

**Comment by mr-ice** [comment]
Thanks for the mention -- I wasn't following this thread until @nickv2002  and @sumpfork pointed it out.

I don't yet have my Intrigue update pack.  I got the base update and started typing to get dividers printed.  

As to splits, I certainly think the splitpile_tag could work.  I've got a concern that I noted going from dividers to wrappers, however.   Some of the wrappers were printing 1 card width (or zero, I can't be sure) (for exampleI noticed this 'Amulet').  This was fixed by adding an explicit count, but I haven't determined why it was not correct.

The mechanism behind splitpile_tag should account for accumulating the number of cards so that wrappers style divider depths will be calculated properly.

**Comment by nickv2002** [comment]
FWIW: @mr-ice links in https://github.com/sumpfork/dominiontabs/issues/73#issuecomment-251478070 show the 2nd Edition Intrigue new/removed cards.

**Comment by mr-ice** [comment]
Thanks, that actually helps a lot, copy paste is better than typing.

I ran into something curious while putting these in -- newlines in description get honored, but in extra they disappear.  I'm able to put in <br> but not sure if that's the right thing to do.

Anyway, I think I've amended the pull request so you can see them all, but it's messy because there was trailing whitespace elsewhere that I removed when I removed the bit added by json.dump().

**Comment by wvoigt** [comment]
@mr-ice  As for "Amulet", it was an error.   Issue #81 already addresses it.  It is covered under pull request #82.  If you find any others please let me know.  I printed out my vertical wrappers a couple of weeks ago, found the errors,  and make the updates.

@nickv2002  Splits today are covered in the card_groups.json file.  As a result, when grouping for wrappers, the total card count will be the sum of all of the cards grouped together.
I think it makes sense to keep this separate from the cards.json.  Individual card information in one place, groupings in one place. 

What I would like to see different in card_groups.json is a unique identifier ("card_tag") for the group, and then change the "new_name" to be "name" and to then also include a "description" and "extra".  That way the grouping (splits in this case) can have their own text for the whole group.  As it is today, it uses the text from the first card. Also, if we ever simplify the translations, these 3 fields could easily be included in the translation files (referenced by the "card_tag")

And in the spirit of keeping like things together,  I propose we should have a sets.json.  So this would have the unique identifier for the set, its name, it's image file, etc, and then a listing for each card ("card_tag") associated with the set.  This would mean the cards.json would no longer need the sets identifier, and we would only need one copy of gold, silver, and any other cards doubled up.  We could even have 2 different versions of cards (with different identifiers).  In this way, 2nd edition sets are just separate sets, that point to many of the same cards and a few different cards.  We correct one card, and it is corrected in all sets it appears in.  So the structure is:

- cards.json describes individual cards.  No need for duplicates.
- set.json describes what cards are in each set.
- groups.json describes card group substitutions.  Take these x cards out and substitute it with the card information given.

Once the json is chagned, then the code changes are mostly in read_write_card_data():
1. load cards.json (as a starting database, not as actual cards as we do today)
2. go through the sets that are requested, adding the cards for each set from the cards.json
3. process all the card groupings
4. process the set.json, groups.json, and cards.json for the language requested. (or maybe we combine these into one file, since everything has a unique identifier now.)

At the end, we have the listing of cards we pass on to the rest of the program.  From there on it is the same as it is today.

**Comment by nickv2002** [comment]
@wvoigt I mostly agree with what you're saying but couldn't we just allow `cardset_tag` to be a list similar to the `types` tag and then eliminate the need for a second `set.json` file?  So if Gold was part of several sets we could just list them there, or we could create 2 version of a different card (i.e. one that got a rules wording change for 2nd edition) and have them belong to different sets as long as the `card_tag` was different.

**Comment by wvoigt** [comment]
@nickv2002 There would still be some set information in other places (setImages, setTextIcons, etc).  But I agree changing the cardset_tag to be a list would also work.

**Comment by nickv2002** [comment]
@sumpfork if you're okay with the changes that @wvoigt and I suggest, I can handle creating the `cards.json` file this week sometime (I'm pretty good with scripting text manipulation etc). @wvoigt if you can provide me an example of how you want `groups.json` to format I can make that too (I have a general idea of how it would work but want to be sure I don't miss anything).  You both are more familiar with the code so if you could take charge of the changes [described here](https://github.com/sumpfork/dominiontabs/issues/73#issuecomment-254081253) that would be great.

**Comment by sumpfork** [comment]
On a cursory review I think I'm good with all of this - it's certainly moving in the right direction and we can revise later as we start playing with it.

**Comment by nickv2002** [comment]
I created a little [Gist showing a draft version](https://gist.github.com/nickv2002/076cf795820454e1b96c60fd31fc2b40) of how I think the new files should look.  `cards_DBv2.json` shows the new format for the list of all the cards with each card unique and only showing once. `cards_en_us.json` shows the translated names, descriptions, and extra text. Finally, `cards.json.converter.py` is the script I used to create these JSON files so you can kind of see my work (and suggest changes as needed).

**Comment by wvoigt** [comment]
@nickv2002 This looks good.  I like it.  Just a few comments:

- I *thought* I read somewhere that Intrigue 2nd Edition was not going to have the base cards.  I may be wrong.  But it is worth double checking.
- Can we have sets for 1st Edition + the new 2nd edition cards?  Normally not important, but if folks use wrappers, then the count would be adjusted for the randomizer cards.
- I know this is personal preference and data is data, so I'm not pushing hard on this one.  But in cards_en_us.json could the order be "name", "description", "extra".

> @wvoigt if you can provide me an example of how you want groups.json to format I can make that too (I have a general idea of how it would work but want to be sure I don't miss anything).

I would say for the common DB, keep it similar to today's card_groups.json, but just change card names to be their card_tag.  Then "new_name" becomes "card_tag".  Then the question is do we have a groups.json for the language that has "name", "description", "extra", or do we just add them to the new  cards_en_us.json.  My vote would be to just add them to the cards_en_us.json.  So all card/groups are in the same language file and are in the same format.

We will need a small sets.json for the set language information.  Example:

```
"intrigue2ndEdition": {
        "set_name": "Intrigue 2nd Edition",
        "text_icon": "I"
}
```

**Comment by nickv2002** [comment]
> - I thought I read somewhere that Intrigue 2nd Edition was not going to have the base cards. I may be wrong. But it is worth double checking.

Good call, [you are correct](http://wiki.dominionstrategy.com/index.php/Intrigue#Basic_Supply_Cards). I will fix this.

> - Can we have sets for 1st Edition + the new 2nd edition cards? Normally not important, but if folks use wrappers, then the count would be adjusted for the randomizer cards.

Yes, that's a good idea. Actually we need a set that's just the new 2ndEd stuff which you can print with the 1stEd set.

> - I know this is personal preference and data is data, so I'm not pushing hard on this one. But in cards_en_us.json could the order be "name", "description", "extra".

I agree and tried to print it that way. However printing the dictionary of translation data proved to be tricky because the line-breaks and characters outside of UTF-8 (I spent an hour trying various thing) so instead I just threw it into `json.dump(en_Translation_cards, open(outTransFile, 'wb'))`. and let that handle everything but lost control of the print order. If it's really important, I can see about fixing it in the final version, maybe just with a `sed` swap or something like that.

I'll also work on the sets and groups JSON info.

**Comment by nickv2002** [comment]
Check out my [updated Gist](https://gist.github.com/nickv2002/076cf795820454e1b96c60fd31fc2b40) for files revised based on @wvoigt suggestions above.

- I fixed the 2nd Edition of Intrigue to not have base cards.
- I fixed the `splitpile_tag` to also apply to Events and Landmarks so they can be combined.
- I added a `sets_en_us.json` for translating set names.

Questions for @wvoigt ?

1. Doesn't the `splitpile_tag` fulfill the _common DB_ you were talking about above?
1. I wasn't sure what the `text_icon` field was specifically for, so I just auto-populated it with the first letter of a set's name.

**Comment by wvoigt** [comment]
@nickv2002 The `splitpile_tag` will make it a little harder to do, and is not quite as easy to "see" what is in each grouping.  But I can see the value of having everything card related in one file.  And what's a little more code, eh?

But can we call it `group_tag` instead?  To me these are cards that are grouped together.  I don't think splitpile describes it well.  And not all of these are categorized as Dominion splitpiles.

A few other comments:
- `card_tag` and  `splitpile_tag/group_tag` need to be unique.  I see "Page" used in both.  The group tag should probably be "Page -> Champion".  Similarly the "Peasant" group should be "Peasant -> Teacher".  I didn't check for others, but we do need to make sure each tag is unique.  All translations are tied to that tag.
- I think it would be best to not use slashes `/` in the tags.  I would substitute a dash.  I guess one question is if we should have any other "rules" for tags to follow.  For example, case, other characters to avoid, etc.
- the `cards_en_us.json` file needs entries for each of the `splitpile_tag/group_tag` entries.

The text_icon is the text version of the set image.  This is the current English version:
```
    'dominion': 'D',
    'intrigue': 'I',
    'seaside': 'S',
    'prosperity': 'P',
    'alchemy': 'A',
    'cornucopia': 'C',
    'cornucopia extras': 'C',
    'hinterlands': 'H',
    'dark ages': 'DA',
    'dark ages extras': 'DA',
    'guilds': 'G',
    'adventures': 'Ad',
    'adventures extras': 'Ad',
    'empires': 'E',
    'empires extras': 'E'
```
Also, while we are adding in everything to `cards_db.json`, how about adding an 
`"image" : "image_file_name",` 
for each of the promo cards:
```
    'walled village' = 'walled_village_set.png',
    'stash' = 'stash_set.png',
    'governor' = 'governor_set.png',
    'black market' = 'black_market_set.png',
    'envoy' = 'envoy_set.png',
    'prince' = 'prince_set.png',
    'summon' = 'summon_set.png'
```
for for example:
```
{
    "card_tag": "Walled Village",
    "cardset_tags": [
        "promo",
    ],
    "cost": "4",
    "image": "walled_village_set.png",
    "types": [
        "Action",
    ],
}
```
Nice work!  This is really starting to take shape.

**Comment by nickv2002** [comment]
@wvoigt Thanks for the feedback. [See updates on the Gist](https://gist.github.com/nickv2002/076cf795820454e1b96c60fd31fc2b40).

I fixed up the `text_icon` tag to match what you listed above.

I added `image` tags to the relevant promo cards. If there's no `image` tag the code should just use the default one for the set it's printed with, which might vary depending on if the card is grouped with a 1st or 2nd edition printout. Also, FWIW, the flag promo image from _Summon_ will be the default promo image from now on (including being used for _Sauna/Avanto_) so I just renamed that image to `promo_set.png`.

> - `card_tag` and `splitpile_tag/group_tag` need to be unique. I see "Page" used in both. The group tag should probably be "Page -> Champion". Similarly the "Peasant" group should be "Peasant -> Teacher". I didn't check for others, but we do need to make sure each tag is unique. All translations are tied to that tag.
Fixed to `Page -> Champion` and `Peasant -> Teacher`
> - I think it would be best to not use slashes / in the tags. I would substitute a dash. I guess one question is if we should have any other "rules" for tags to follow. For example, case, other characters to avoid, etc.
Fixed to be `-`
> - the `cards_en_us.json` file needs entries for each of the `splitpile_tag/group_tag` entries.
Not sure what specifically you meant here? They should all be there now. Is there any you see missing?

You're correct that not having a separate groups file will be a bit more coding, but it shouldn't be too bad, just iterate through the entire Cards_DB after reading it in and build a reversed dictionary keyed by each set or group. I can write that code loop if it helps. Seems like we've essentially building a Many-to-Many relationship between sets and cards.

Thanks!

**Comment by wvoigt** [comment]
@nickv2002 Good work.
Yes, that was my thoughts on the `image ` tag.  Use the set's image unless a card has a specified one.

> _the cards_en_us.json file needs entries for each of the splitpile_tag/group_tag entries._
> Not sure what specifically you meant here? They should all be there now. Is there any you see missing?

For example, there should now be entries in `cards_en_us.json` for:
- `Page -> Champion`
- `Peasant -> Teacher`
- `Sauna - Avanto`
- ... and any other unique tag under `group_tag`
That way the card for the group can have a different name, description, and extra from any other card.

Oh, just found another one.  `Tournament`  is used as a `card_tag` and `group_tag`.  The group_tag should probably be something like `Tournament and Prizes` if it includes the "Tournament" card, or just `Prizes` if not. 

I'm fine with using a single Cards_DB file.  As you said, the code is not too bad.  I haven't had time to actually code everything up to test.   But I've worked out all the pieces on paper.  I just need some keyboard time.  Hopefully this weekend.

**Comment by nickv2002** [comment]
@wvoigt Thanks for all the feedback.  I had a busy day but I'll fix the Tournament to group with the Prizes (`Tournaments and Prizes`) this weekend. 👍

**Comment by wvoigt** [comment]
@nickv2002 I just noticed that `debtcost` is not making it over into the new DB on the cards that cost debt.

**Comment by nickv2002** [comment]
@wvoigt thanks for the corrections. I just uploaded a [new version](https://gist.github.com/nickv2002/076cf795820454e1b96c60fd31fc2b40/raw/48c97931f10f3f6d1714b2dfd9774b8386ccc41a/cards_db.json) with a group_tag for `Tournament and Prizes` and the `debtcost` tag.

**Comment by sumpfork** [comment]
Starting to look great. Do you want me to merge (or do you want to merge) #83 as a deployable though inferior interim solution for those with the new base cards, or rather wait for the reworked proper solution? Sorry I'm contributing little - 6 months old at home, formed my own company, that kind of thing...

**Comment by nickv2002** [comment]
@wvoigt how's the integration coming? Do you need help with this this week? Trying to establish an ETA to answer @sumpfork's question…

**Comment by wvoigt** [comment]
@nickv2002 Sorry, worked on it this weekend, but forgot to post.  I'm finished with the initial implementation.  Not saying it is perfect (I'm sure there are bugs), but I think it is far enough along for others to tryout.  Here is a [link on Gist](https://gist.github.com/wvoigt/28dde067244b18a068dd087073a1dd10) to the files.

How should we go about combining all of this?  Do you want to create a branch that we then both work on?  Then when we are done, the branch gets submitted?

We also need to start thinking about how we create the text translation files for the other currently supported languages (de, fr, it).

**Implementation Notes:**

The `cards_db.json` is currently not a valid JSON file.  A handy tool is at http://jsonlint.com.  The json load errored on the extra commas at the end of lists. Similarly the sets text file has a trailing comma for the last entry.

If `--special_card_groups` is used and the resulting `group_tag` does not have any language translation (even in 'en_us'), it will print an error on the card similar to:
_ERROR: Missing language entry for group_tab 'Page -> Champion'._
This should really only be for us, to help find entries that need to be made in the 'en_us' translation file.
After that, then the English text will be there to remind others about their own translation.  This allows for a simple search for "ERROR" in the PDF file.

I added an option called `--edition` that can be "1", "2", "latest", or "all".  Defaults to "all".
This allows for a quick filtering for those that don't want everything (i.e., "all").
- "1" is for all 1st editions of expansions.  It also includes 2nd edition update packs.
- "2" is for all 2nd editions of expansions.  So base, Dominion 2nd ed., and Intrigue 2nd ed. (no update packs, since already in 2nd edition)
- "latest" is for base, Dominion 2nd ed., and Intrigue 2nd ed., and all the remaining 1st editions.

I created a `sets_db.json` to hold additional set information including the `"image"` file, default `"set_name"` and "`text_icon"` values, and the `"edition"` information for the above option.
The default `"set_name"` is `"*<cardset_tag>*"`.  Again to help point out if we are missing something in the default translation file.  This had the side effect of removing a lot of data from cards.py.
    
I assumed the following file structure (I'm open to suggested changes):
```
card_db/
    cards_db.json
    sets_db.json
    <language>/
        cards_text.json
        sets_text.json
```
This had the advantage of not reusing any existing file names.  And I think they are descriptive.

I have not changed the `dray.py` file yet. I was limiting my updates for now.  But if the Mixed case for the page footer bothers anyone, the change is on **line 797**:
Change:
    `setTitle = c.cardset.title()`
to:
    `setTitle = c.cardset`
    
**Question**
With Events, Landmarks, and Prizes now included in --special_card_groups, I've been trying to figure out what should become of 
`--exclude_events`, `--exclude_landmarks`, and `--exclude_prizes`.
One option is to just remove the options.  Another would be to change their meaning.
For example, `--exclude_events` could be changed to group all the events across all expansions.
Similarly for Landmarks (if they show up in other expansions).
The question is: where to put these?  They span the whole set of expansions now.  Maybe another "Extras" section at the very back?

The `--exclude_prizes` could morph into an option that groups the prizes together, but not combine with "Tournament".  Or, we go the opposite, we change the grouping to be "Prizes" instead of "Tournament and Prizes", and then --exclude_prizes groups them with "Tournament".

In any case, that part of the code is questionable at this point.

Let me know if you find anything.

**Comment by nickv2002** [comment]
@wvoigt Thanks! all that sounds great. I'd say to go ahead and make a new branch.

> The cards_db.json is currently not a valid JSON file.

Yeah, I was just manually removing those training commas rather than write some smart code to skip the last ones and forgot to do it the last time I updated.  Updated to remove the commas now.

> We also need to start thinking about how we create the text translation files for the other currently supported languages (de, fr, it).

I can adapt my scripts to create language files based on the translations we've currently got, but let's get the English version working first.

I like your `--edition` filter.  Great idea.

For `--exclude_*` let's just focus on 2 basic modes: all the cards (nothing combined and grouped) as the default, and a _max grouping_ mode where everything that is combine-able (prizes, travelers, events, landmarks, split piles, etc) is put together, even if it's across sets.  I think that should satisfy most people (and if you want separate Events for each set they're in, just create 2 PDFs, one for each set).

For, all your other questions I think your instincts are fine.

**Comment by mr-ice** [comment]
> A handy tool is at http://jsonlint.com.

I can also recommend the node json tool at https://www.npmjs.com/package/json

alternately, use the json only through the python json module, so you're just dealing with data structures in one language.

**Comment by wvoigt** [comment]
> alternately, use the json only through the python json module, so you're just dealing with data structures in one language.

@mr-ice , I don't understand the comment.  What do you mean by one language?  The data is loaded through the json module.

**Comment by nickv2002** [comment]
@mr-ice I get what you're saying, but the library wasn't giving me control on the _order_ that the arguments were written out for each dictionary item so I just wrote plain text.

**Comment by nickv2002** [comment]
Sorry, hit **Close & Comment** instead of just **Comment**

**Comment by wvoigt** [comment]
> For --exclude_* let's just focus on 2 basic modes: all the cards (nothing combined and grouped) as the default, and a max grouping mode where everything that is combine-able (prizes, travelers, events, landmarks, split piles, etc) is put together, even if it's across sets. I think that should satisfy most people (and if you want separate Events for each set they're in, just create 2 PDFs, one for each set).

I think I'm partial to 3 groupings:

- Default to nothing combined/grouped
- In expansion grouping invoked with `--special_card_groups` as now implemented.
- Max grouping across expansions. Currently I think this only applies to Events. Any other cross expansion groups?

To do the Max grouping there would need to be some dummy expansion set to hold the grouped cards. Not sure what it should be called or what the set image should be.

**Comment by wvoigt** [comment]
@nickv2002 , 

> I'd say to go ahead and make a new branch.

I guess I was looking for some guidance.  Do I make the branch under sumpfork/dominiontabs (can I even do that?), under you, or under me?  Does anyone of them have an advantage for us in making our individual changes?

> I can adapt my scripts to create language files based on the translations we've currently got, but let's get the English version working first.

I agree on the English version.  I think we are very close there.  Mostly missing group entries.
What I fear is that the mapping may be tricky unless you know the language.  There will need to be a mapping for each English `"card_tag"` to each card "name" for the language.  Just worried there will be a lot of mapping by hand since the way the files were structured, everything including the `"name"` was in the target language.

**Comment by mr-ice** [comment]
> What do you mean by one language?

Meaning you can use the json python library to write your json then you manipulate the data structure in python -- you can do this interactively or through scriptlets.  Never writing anything in json by hand.   The drawback seems to be what @nickv2002 mentioned:

> library wasn't giving me control on the order that the arguments were written out for each dictionary item so I just wrote plain text.

which is a perfectly fair comment, the library only does "hash" order and "sorted" order.

It seems like sorted order was the way I found it when I contributed my tiny bit.  Is it reasonable to pick one of those (maybe 'sorted' since it's at  least predictable)?

example: [Gist](https://gist.github.com/mr-ice/5579cd94956150080536e8e99eebfcdb)

**Comment by mr-ice** [comment]
Alternately, you can use an OrderedDict like this
```
from collections import OrderedDict
data = json.load(file('cards.json'), object_pairs_hook=OrderedDict)
```

Which will preserve the order you find.  nothing different required on json.dump().  New keys have to be appended.

New [Gist](https://gist.github.com/mr-ice/a693375f4a05b4e6e19ebead27948218)

**Comment by nickv2002** [comment]
@mr-ice yeah an OrderedDict is probably correct way to do it, but I just dug in and wrote the files myself for expediency.  Thanks for the tip though.

**Comment by nickv2002** [comment]
@wvoigt We're both _Collaborators_ now so we should have all the privileges to create/merge branches. I don't know if there a most-agreed-upon-correct way to do branches and merges, so I think it makes sense for you to create a new branch here with your work and I can help contribute to that.  Once we're happy we can merge back to the master branch.

As for the non-English language stuff, I just looked at the files and it's not hard to figure out what text is what for most of the cards based on description/cost/set. Maybe I can write a script to create the JSON file with all the possible English names for each entry so it's easy to update by just deleting the wrong text, then I can go through and make a quick first pass and finally we can solicit help from some folks on BGG for what remains. It's a good point to update everything too, now that we've got the new system, since many of the translations are out of date.

**Comment by wvoigt** [comment]
@nickv2002 I created a branch called ["Card-DB-refactoring"](https://github.com/sumpfork/dominiontabs/tree/Card-DB-refactoring) to hold the changes. Please let me know if I should have done something differently.  At the moment, all test fail since all the parts are not there yet.

So I'm looking at the new generated dividers, and in particular the expansion dividers.  And while we are focused on "1st Edition" vs "2nd Edition" currently, ultimately when I print my dividers I don't think I really want to see "**Dominion 1st Edition**" and "**Intrigue 2nd Edition**". I think I would rather see just "**Dominion**" and "**Intrigue**". I know what version they are.  Now I know others would probably want to see the edition information.  So the question is how to provide both.

I propose that the `sets_db.json` file be changed slightly to include an optional` "short_name"` field that contains a shorter name.  For example
```
    "dominion1stEdition": {
        "image": "base_set.png",
        "set_name": "*dominion1stEdition*",
        "short_name": "Dominion",
        "text_icon": "",
        "edition": [
            "1"
        ]
    },
```
Then with an option we can either assign just the `set_name` as before, or the `short_name` if it exists. I'm open as to which would be the default.  `"short_name"` could also be included in the `sets_text.json` file for each expansion that might need a shorter name in a different language.  
```
    "dominion1stEdition": {
        "set_name": "Dominion 1st Edition",
        "short_name": "Dominion",
        "text_icon": "D1"
    },
```
And again, this is only for the text on the expansion divider.  The set name on the footer would still show the longer name.

In other updates:

I added an option called `--upgrade_with_expansion` which will put the upgraded cards into the corresponding earlier expansion.  So all 1st edition cards as well as the upgrade cards appear in the 1st edition set.   That way the new cards are listed on the expansion dividers and any tab/ordering fit the expansion as a whole.  And because cards were deleted in the 2nd editions, this combined set of cards is different from the 2nd edition.

I fixed `--exclude_events` to be the max version that we discussed.  The resulting group was put into an "Extras" set.  I did the same for `--exclude_landmarks`, but I know this really doesn't cross expansions at this time.  It was more for keeping the existing option.  At the moment `--exclude_prizes` does not do anything.

**Comment by nickv2002** [comment]
@wvoigt Great work. I think you did the branch right, I just created and merged a pull request on that branch to add updated set icons & the new format JSON DB files. https://github.com/sumpfork/dominiontabs/pull/87

New icons are from [this BGG file](https://boardgamegeek.com/filepage/73273/dominion-card-icons-vector-images). Notably, the first and second editions of both Dominion and Intrigue have slightly different icons, so I remapped them using the new `sets_db.json`.

I think defaulting to the `short_name` field is fine.  I don't use the Expansion Dividers, so I don't even see set name on those once they're cut out.  I'll add `short_name` to the `sets_en_us.json` file in my next update.

`--upgrade_with_expansion` is a good option.  Just make sure the new cards still get the newer set icon that's linked from their set in `sets_db.json`

Questions:

1. In `sets_db.json` do you actually want to fill in the `"text_icon": "",` line? They're all blank now so why have them?
2. In `sets_db.json` why all the `*` on both sides of the `set_name` tags? (Except on Promos: `"set_name": "*promo",`) Are they supposed to be everywhere? Again, why have them if they're everywhere?

**Comment by sumpfork** [comment]
Great work folks, let me know when you're close enough for me to start testing and reviewing.

**Comment by wvoigt** [comment]
@nickv2002 
> In sets_db.json why all the * on both sides of the set_name tags? (Except on Promos: "set_name": "*promo",) Are they supposed to be everywhere? Again, why have them if they're everywhere?

The intent was two fold:

1. Have at least some default value for each item
2. Have some visible indication on the divider that the main default language was missing a translation.

This branch should be mostly functional now.  We are missing some translations, but I believe all the other parts are there.   Of course it fails all the continuous testing, so changes will need to be made there.

I updated the branch with the following:

-  Updated `__init__.py` to include` --expansion_dividers_long_name` option;  fixed the set image when combining upgrade sets; fixed a few bugs;  and changed the file structure to:
```
card_db/
    cards_db.json
    sets_db.json
    <language>/
        cards_<language>.json
        sets_<language>.json
```

-  Updated `sets_db.json` to include `"*"` for default "text_icon" and fixed `"set_name": "*promo*"`
-  Added back `inimages/base_set.png` since it was being called out for the base set. At least until a better image can be substituted.
-  Updated `cards_db.json` to fix json errors.

**Comment by nickv2002** [comment]
Good work. Thanks.

> -  Added back `images/base_set.png` since it was being called out for the base set. At least until a better image can be substituted.

But the base cards should literally not have an expansion symbol.  [See these scans.](http://dominion.diehrstraits.com/?set=Base%20Cards). They work with everything, they're in every game, it's just the Dominion starter set that gets a copy of them since they don't want you to have to buy 2 boxes to start playing.

**Comment by wvoigt** [comment]
@nickv2002 
> But the base cards should literally not have an expansion symbol.

You are right.  Now fixed in the branch.  `base_set.png` is removed (again!)

Looking at the output, it looks like we are only missing translation entries in `cards_en_us.json` for the following groupings:
- 'Page -> Champion'
- 'Peasant -> Teacher'
- 'adventures events'
- 'Tournament and Prizes'
- 'Hermit - Madman'
- 'Shelters'
- 'Urchin - Mercenary'
- 'Bustling Village - Settlers'
- 'Catapult - Rocks'
- 'Encampment - Plunder'
- 'Gladiator - Fortune'
- 'Patrician - Emporium'
- 'empires landmarks'
- 'empires events'
- 'Sauna - Avanto'
- 'promo events'
- 'events'
- 'landmarks'

and in `sets_en_us.json` we are missing:

- 'extras'

**Comment by mr-ice** [comment]
> I noticed the potion image looks a little weird on the tabs

It looks like it is being stretched.  It may need to be padded width wise.  I might suggest that anyway -- make all the icons the same size so that they will be predictable when placing.

> missing translation entries

Should some of these be all lowercase?

I don't understand something here:

'landmarks' vs 'empires landmarks'

First, are these (and others) supposed to be all lowercase?  (maybe just counter-intuitive or I haven't been paying enough attention to note 'cards', vs 'sets', vs 'groups', vs 'stacks' vs. 'card_tag')

Second, is the intention to have these cards be in both groups, so that if landmarks are added in a future set they can either be printed together -- such as 'landmarks' all together or separately 'empires landmarks' and '* landmarks' etc.?

Finally, I've done a quick commandline python [Gist](https://gist.github.com/mr-ice/2df968679155fe074a0570a298091a35)  to get the contents of the ones I thought I understood moved into the new cards_en_us.json format.  Note the checkout directories I mention in the title of that gist (you need to have both branches checked out or change the paths).

Looks like something similar could be done with Page -> and Peasant -> .   Not sure about the other missing ones.

**Comment by nickv2002** [comment]
I missed the graphical issues with the new set icons so I'll try and square them all off to the same size later today.

@mr-ice thanks for you work on the recombination. Please submit a pull request when you get a chance.

**Comment by wvoigt** [comment]
> It looks like it is being stretched. It may need to be padded width wise. I might suggest that anyway -- make all the icons the same size so that they will be predictable when placing.

Actually, when printed on the tab, it looks compressed horizontally to me.  I agree it makes sense to have the images the same size for the same use.  Coins/Potion/Debt the same, set images the same.

> Should some of these be all lowercase?

I listed them as they appear in the group_tag.  Some are in lower case, others in mixed case.
Personally, I think it might be better to keep all *_tag entries as lower case.  Then it is easier to see if we are talking a tag or a name. But either can work.  The code just needs them to be unique.

> 'landmarks' vs 'empires landmarks'

Yes, thinking ahead.  Just like Events.  The group_tag will group inside an expansion.  And then an option to combine across expansions into a single group. And that group will be placed in an expansion/set called "Extras".  I know it really doesn't apply yet to Landmarks, but they are just ripe for additions into promos and other expansions.  I would think the extra and description text for each would be the same.  Just as it can be the same for each of the Event groups.

> Not sure about the other missing ones.

'Page -> Champion' and 'Peasant -> Teacher' are the only ones that don't have translations today.  So I would think you can easily pick up the rest.

**_In other thoughts..._**
Right now, as cards are being combined into a group, the very first card found is used as the holder for that group.  What that means the group uses that card's cost, debtcost, potcost.  Now for Events and Landmarks, since they have their own card type, the code can check and zero out those values.  But for the others, we want to have the cost of the first/primary card.  So order is important in the cards_db.json file.  One way to bullet proof it would be to have a separate groups_db.json file that contained the cost information for the group.  Not sure it is worth it or not.

**Comment by nickv2002** [comment]
merged https://github.com/sumpfork/dominiontabs/pull/91 to make all set images 300x300 for consistency.  Should fix print problems but haven't tested yet

**Comment by nickv2002** [comment]
> Right now, as cards are being combined into a group, the very first card found is used as the holder for that group. What that means the group uses that card's cost, debtcost, potcost. Now for Events and Landmarks, since they have their own card type, the code can check and zero out those values. But for the others, we want to have the cost of the first/primary card. So order is important in the cards_db.json file. One way to bullet proof it would be to have a separate groups_db.json file that contained the cost information for the group. Not sure it is worth it or not.

Grouped cards should be sorted by cost generally, because (at least so far) all piles and such put the cheaper stuff on top (makes sense for the game because you can't afford expensive stuff early). I would rather not maintain a separate `groups_db.json` file if we can avoid it.

**Comment by wvoigt** [comment]
@nickv2002 
The new images look good.  Potion (as a cost) is much better now.

> Grouped cards should be sorted by cost generally

I did not realize that.  Updated the code to always update to the lowest card cost.  Cards that have a cost that includes a "*" will treat that cost as a large number (so should never be the lowest card cost).

I also fixed up some group_tag errors in `cards_db.json`, and added all the missing text entries in `cards_en_us.json` and `sets_en_us.json`. 

`'Page -> Champion'` and` 'Peasant -> Teacher'` still need real text to be added.  But it just got too late tonight.

From my standpoint, the branch is pretty functional right now.   I think the remaining TODO's are:
- Finish text for `'Page -> Champion'` and` 'Peasant -> Teacher'`
- Review all text to make sure it is accurate
- Create the language files for de, it, and fr to at least the same level as exists with today's release.
- Find and correct the continuous integration failures
- Test, Test, and Test some more :-)

**Comment by wvoigt** [comment]
I  added text for `'Page -> Champion'` and` 'Peasant -> Teacher'`
I corrected the Travis continuous integration failures.

@nickv2002  I think we are down to the final steps:
- Create the language files for de, it, and fr to at least the same level as exists with today's release.
- Review all text, cards images, options, etc
- Test, Test, and Test some more :-)

@sumpfork I think it is close enough for you to start testing and reviewing.  A lot of it can be done without the other language files.

**Comment by nickv2002** [comment]
Nice work @wvoigt 👍

Busy this weekend but I can poke at the translations this week.  We should not hold up the english update for the translations however, as it will take more than my work to get them up-to-date.

**Comment by wvoigt** [comment]
@nickv2002  I'm not sure what is going on now.  I did a clean start off a download of the current branch.  Now I'm seeing problems with the following images:

- potion_small.png - horizontally compressed (maybe?) and not a clear background
- alchemy_set.png - black box
- dominion1StEdition_set.png - black box

Any idea on what could be causing that?  The others seem to be OK.

![potion_small and alchemy_set](https://cloud.githubusercontent.com/assets/1182468/19832446/565cd044-9de9-11e6-90b7-411cedd30c47.png)
![dominion1stedition_set](https://cloud.githubusercontent.com/assets/1182468/19832447/593beea8-9de9-11e6-8bc5-fd9fa85c5ed1.png)

**Comment by nickv2002** [comment]
@wvoigt no idea, haven't changed anything since I changed all the set icons to be 300x300 pixels but that didn't change the colored potion.

Perhaps related: should it be `dominion1StEdition_set.png` or dominion1stEdition_set.png`? looks like the file is `dominion1StEdition_set.png` but I think it should probably be `dominion1stEdition_set.png`

**Comment by wvoigt** [comment]
@nickv2002 In order to help speed up the creation of the language translations from the existing cards.json for each language, I've created a cross reference spreadsheet.  It would not let me attach it here (even if I made it .zip).  So I added it for now in  `\tools\CrossReference.ods` (we can delete it later).   Names that I found in the `cards.json` files are highlighted in blue.  Many are missing.  I found a [list of cards across various languages](http://wiki.dominionstrategy.com/index.php/List_of_cards_in_other_languages) that I used to fill out the spreadsheet.  I tweaked a few to match the name used in the `cards.json` file.

I would suggest we keep the same order of cards in each language file the same as in `cards_en_us.json`.
If the language's cards.json has the card listed, then provide all the information that is not empty (name, description, extra).   I noticed some translations just put empty strings for the description or extra.  I think we should skip those so that the English version shows.  That will hopefully prompt others to help translate.

If an name does not exist, then provide the name if we have it in the spreadsheet.
```
	"Fountain": {
		"name": "Brunnen"
	},
```
And if we don't have that, then a blank/holder record for that card:
```
	"Fountain": {
	},
```
That should make it easier for future translations.  Then in the future, we add any new cards to the bottom of the file.  And for each card that is not complete, the information can be added as it is provided.

What do you think?

Now back to looking at the images again...

**Comment by nickv2002** [comment]
@wvoigt Your proposal sounds fine but wouldn't it be more useful to put then English info in if the translated info wasn't available so it could be translated without referring to a second file?

Busy with other stuff so haven't got to work on this much but I have written some scripts that can combine all the names in the spreadsheet you uploaded into a `cards_<language>.json` file.

**Comment by wvoigt** [comment]
@nickv2002 I'm of two minds on having the English in the other language files:
1. Yes, it makes it easier to do the translation in place, using a single file.  +1 for translation.
2. Are we going to remember to always move over the English if card wording changes (as it just did with some of the 2nd edition cards).  And are we going to remember to add the English to each other language file when a new card comes out. +1 for maintenance to leave out the English.
I can go either way on that.

In other work, I just made updates that hopefully will make the comparison easier.  The files had two different encodings.  Parts used \uxxxx (unicode) and parts just had the ISO8859 characters.  And the two don't compare very easily. Even the English file had a few backslash characters that could be removed.  I is hard to read in the files without getting double encoding (\\uxxxx, and \\/).  So I just standardized on one encoding.  When the files are read in, just make sure they are encoded from ISO8859-15 (really ISO8859-1 works too except for one French character.)  I also uploaded a csv file to make it easier to directly read from python.

Also, I noticed that except for the card names, the French file was just in English.  And it did not have as many cards.  So I would suggest just using the names from the spreadsheet.

And as a consequence of going through all the cards, I noticed that we have lumped all the Ruins together.  While the German version had them separated out.   So given some time, it might be good to add the 5 individual Ruins cards and then place them all in the Ruins group.  Then there is flexibility.  Not critical for now, but just wanted to mention it here before I forget.

**Comment by nickv2002** [comment]
@wvoigt Thanks. I tweaked a few more things including removing some unneeded unicode in the card DB for: https://github.com/sumpfork/dominiontabs/pull/99

For card translation DBs how about option 3: Put in English wording for now, as part of the large transition, but not in the future.  I can also add an `untranslated` tag to each card that needs it when my script merges everything together.

**Comment by wvoigt** [comment]
@nickv2002 

> For card translation DBs how about option 3: Put in English wording for now, as part of the large transition, but not in the future. I can also add an untranslated tag to each card that needs it when my script merges everything together.

I like this option.  Especially if the field can include the names of the fields that are not translated.  This allows the option in the future to have a maintenance script that can automatically update if needed.  The maintenance script could also add any missing cards, make sure all entries are ordered the same, etc.

> As for Ruins I think it's correct not to break them out separately. They are always 1 pile that should be in a random order. There is no aspect to the game that refers to them separately. They're really 1 card that has a few different effects.

While I see where you are coming from for this, I fail to see how it is any different than many of the other "groups" that we have - Events, Landmarks, Sauna - Avanto, etc.  While some are random and others are not, the point is they all have different card names but their use are tightly related.  Some may like them combined for storage, others may not.  Now personally, I would store them together.   

My 2nd argument for adding the 5 cards (and then grouping) would be that the 5 cards are currently in the German translation - so the German dividers currently print these 5 cards.  I was hoping to keep this update so that no one lost anything.  And by using the group, really folks can do it whichever way they want.

**Comment by wvoigt** [comment]
@nickv2002 
On the image issues, I'm not sure if this is it, but see the following changes:
- `alchemy_set.png` new version has bit depth is 32, working set images are 8.  But old ones are 32 as well.  I suspect some change in the transparency.   But I'm not an image expert.
- `dominion1stEdition_set.png`  same as above
- `potion.png`  It is the same size as before, but older version had 32 bit depth, current one only 8. 
   **Note**: I mistakenly called out the wrong image on this before.  Upon re-examination, it is not so much compressed, it is just that the background is not clear.

At this point I'm stuck.

**Comment by nickv2002** [comment]
@wvoigt hmm, don't have time to test now but my first thought is to just put the old images back in place of the new ones (renaming as needed). Will let you know if I figure out anything else.

**Comment by nickv2002** [comment]
@wvoigt I have the same issues with images in my version but haven't had much time to test anything. Hope to soon but work has been busy and might still be...

**Comment by nickv2002** [comment]
@wvoigt images should be fixed from https://github.com/sumpfork/dominiontabs/commit/478296f8b5b7719f9574eca29e4e5151cd9d3527 and https://github.com/sumpfork/dominiontabs/commit/db76fffdff84d7019c152249d6ba0d11387adc11

Give that a try and see if it works for you too.

**Comment by soudek01** [comment]
Hey guys, I know nothing about programming, but I love your dominion tabs! Seriously, thank you all so much! Any guesses about when the V2 cards will be up on http://domtabs.sandflea.org/?

**Comment by sumpfork** [comment]
When @nickv2002 and @wvoigt believe the code V2 is ready they'll make a pull request against master and I'll review the changes - despite being busy I promise to try to be quick on this, as I know there are people waiting. Once that process is done I'll put them up at the online generator, but there'll be a bit of work involved in that as some of the options are changing.

**Comment by soudek01** [comment]
Awesome,

You guys are the best!

**Comment by nickv2002** [comment]
I think it's basically ready in English.  Not ready for other languages.  Will leave final say up to @wvoigt because he's more familiar with the detailed bits of the codebase than me, but my test output looks correct in a few checks.

**Comment by wvoigt** [comment]
I believe the code base is done.  I have one small change I want to make, but it is not critical.  
Right now, more eyes looking at the output would help the most, just to make us feel more confident we have not missed something along the way.

I'm busy for the next two days, but I can work on it this weekend to finish everything up.

@sumpfork , I'll also post a summary of all option changes here.

**Comment by wvoigt** [comment]
@nickv2002  I'm almost finished with the translations [de, it, fr].   Before I post the changes, I thought I would state what I have done,  just in case I went a little too far.

- In the language files, if elements of a card translation do not exist, the en_us version is assigned.
- Implemented the `untranslated` field to help keep track of what elements have been translated for a card.  Basically, it is a string with the field names "name", "description", "extra" to indicate the elements that have not been translated.  As elements are translated, the name of the element should be removed from the list.
- Created a `translation.txt` file that will reside in each language directory with instructions on how to update the files.
- Created a language directory `xx` that has a starting point for any new translation.  This differs from the en_us file since each card will have the  `untranslated` field set.  The en_us files, by definition have been translated and do not need the `untranslated` field.
- in the main program, the allowed languages are now automatically calculated.  So adding a new language is as simple as adding it's directory with the proper files.

I've wrote a maintenance tool called `update_language.py` for the tools directory that will:

- Sort the cards in `cards_db.json` by `cardset_tags` and then `card_tag`.  Before the main sort, the items in `cardset_tags` are first sorted (with 'base' going in front if it exists in the tags).
- Updates each language directory, adding any new cards and sets found.  In addition each `cards_XX.json` is put in the same order as the newly sorted `cards_db.json` file.  This should make finding cards easier.  Any entries no longer found in `cards_db.json` are kept at the end, just in case they are needed in the future.
- Empty `untranslated` entries are removed since the card is fully translated.
- Similar updates for sets. Sort the sets in `sets_db.json` by the `cardset_tags`.  Again, each sets_XX.json is put in the same order.  Any new set information is added to the file.
- Updates are to a new `card_db` directory under the tools.  So the results must be copied by hand.  That way the files are not accidentally harmed.  And the results can be checked before updating.


The intent was a utility that could be run any time a card/set is added or updated that would propagate the update to all the languages.  At that point, translations in each language can start.

If there are no objections, I'll post the changes tomorrow.

**Comment by nickv2002** [comment]
@wvoigt awesome work, much more comprehensive than originally planned.  Sorry I haven't had time to work on more but work got in the way.  Post what you've got for sure.

**Comment by wvoigt** [comment]
@sumpfork I believe everything is ready to go.  We could use some more folks testing. Here is a summary of the option changes:

**New Options:**
- `--edition`  Select specific editions of sets. This can be combined with other options to refine the expansions to be include in the output.   `--expansions `can still be used to limit the selections even more.  Possible values are:
  '1' is for only 1st Editions
  '2' is for only 2nd Editions
  'latest' is for the latest edition for each expansion
  'all' is for all editions of expansions (default)

- `--upgrade_with_expansion`  Include any new edition upgrade cards with the upgraded expansion.  For example, each of the recent update packs would be included with the 1st edition.  Dividers from cards in the update packs will still print out the symbol for the 2nd edition on the tab to help differentiate the card.
  
- `--expansion_dividers_long_name`  Use the long name with edition information on the expansion divider tab.  Without this, just the expansion name is used on the expansion divider tab.  For example, "Dominion 2nd Edition" instead of the short name "Dominion"

**Changes to existing options:**
- `--special_card_groups`  For the most part this is the same.  The exceptions are that "Tournament and Prizes" is now a group, and Events and Landmarks are each a group inside their expansion.

- `--exclude_events` Has a similar meaning as before, and differs from the Event groupings in `--special_card_groups`  by grouping all Events across all expansions/promos into a single divider that is placed in a special "Extras" set.

- `--exclude_landmarks` Similar behavior  as `--exclude_events` but for Landmarks.  Now currently, all Landmarks are in one set.  But this was left in place in anticipation of more landmarks being added in future sets or promos.

- `--exclude_prizes`  Currently has no meaning.  The prizes are now grouped with Tournament in `--special_card_groups`

**Comment by soudek01** [comment]
Hey guys. Again, fantastic work. I just generated some tabs (you can see
the setting I selected in the attached image) and noticed a few thing to
consider:

1. Possession now has an erratum that the possessor gains all tolkens as
well. See the new rules document at
https://boardgamegeek.com/filepage/131854/complete-rules-dominion-and-all-its-expansions
.
2. Ruins card card count say 10.
3. Moneylender, Throne Room, and Mine should say "you may." (I know this is
a V2 thing that you're probably working on.)
4. For the card count on split piles you could so something like 5/5.
5. Settler/Bustling Village card count says 20.
6. The Adventures' Events has no card count. Empires' Events card count
says 34.
7. Mill is different from all the other cards. It says "worth 1 VP" instead
of a "1" and then a victory point symbol. Also, the dividing line on Mill
didn't generate correctly.
8. V1 and V2 cards generated with the same expansion symbol.
9. Sauna/Avanto card count says 30.

Other Ideas:

1. You could consider including the card type(s) on the bottom of the
"card" tabs.



Here is a link to the tabs I generated:
file:///C:/Users/Sam%20and%20Megan/Downloads/sumpfork_dominion_tabs.pdf.
Thanks for all of your hard work. I'll generate some more tabs in the
future and give you more feedback if you'd like?

Sam Oudekerk

**Comment by wvoigt** [comment]
@soudek01 This is exactly what we need right now.  More eyes checking everything.

I gather the cards you reviewed were at the [website](http://domtabs.sandflea.org/).  If you are able, the code here on GitHub is ahead of what you were able to run.  So some of these items have already been fixed and working their way to the website and other items have not.

>1 Possession now has an erratum that the possessor gains all tolkens as well. 

This needs looked at.  Thank you for the link. Ironically, we have been so focused on code changes, I don't think we have fixed the text yet.

>2 Ruins card card count say 10. 

This has been fixed and is in queue.

>3 Moneylender, Throne Room, and Mine should say "you may." (I know this is a V2 thing that you're probably working on.) 

This needs looked at.  I don't think we have fixed the text yet.

>4 For the card count on split piles you could so something like 5/5. 

I will add this as a feature enhancement.

>5 Settler/Bustling Village card count says 20. 

This has been fixed and is in queue.

>6 The Adventures' Events has no card count. Empires' Events card count says 34. 

This has been fixed and is in queue.

>7 Mill is different from all the other cards. It says "worth 1 VP" instead of a "1" and then a victory point symbol. Also, the dividing line on Mill didn't generate correctly. 

Good catch!  We need to correct that.

>8. V1 and V2 cards generated with the same expansion symbol. 

This has been fixed and is in queue.

>9 Sauna/Avanto card count says 30. 

This has been fixed and is in queue.

>Other Ideas:  You could consider including the card type(s) on the bottom of the "card" tabs.

I will add this as a feature enhancement.   Instead of the bottom, how about the first line at the body top?  The same line as the card counts (and coins and sets icons if selected for body top)

Again, thank you for reviewing the cards and yes, we want the feedback.

**Comment by wvoigt** [comment]
We seem to have lost Possession 2/2 somewhere along the way.

**Comment by soudek01** [comment]
Ok ya'll, here are some more noticings:

1. Distant Lands says "4 VP" and "0 VP" instead of 4 {victory symbol} and 0
{victory symbol}.
2. I am not sure about this, but it seems wierd to me that Page ->Champion
and Peasant -> Teacher just say what Page and Peasant do. Maybe instead do
the description from the Adventures rulebook:

**Comment by soudek01** [comment]
"Adventures has Traveller cards. These cards have an arrow over the text
box to remind players of their ability to upgrade into another card. When a
player discards a Traveller from play, he may exchange it for the card
indicated; he returns the card being exchanged to its pile, takes the card
he is exchanging it for, and puts that card into his discard pile. For
example when exchanging Peasant for Soldier, he puts Peasant back into its
pile and takes a Soldier and puts it into his discard pile. Exchanging is
not trashing or gaining, and so does not trigger abilities like Travelling
Fair's. It is 5 Adventures_Puerto rule us 2/6/15 3:03 PM Page 6 6 optional.
It only happens when the card is discarded from play; discarding it from
hand, such as due to not playing it, will not trigger it. It only happens
if the card being exchanged for has any copies available; if there are no
Soldiers in the pile, Peasant cannot be exchanged at that time. If multiple
cards do something when discarded from play, the player picks the order;
for example, with no Soldiers left in the pile, a player with Peasant and
Soldier in play could first exchange Soldier for Fugitive, then exchange
Peasant for that Soldier. Page and Peasant are Kingdom cards that are
Travellers. Page is exchanged for Treasure Hunter, which is exchanged for
Warrior, which is exchanged for Hero, which is exchanged for Champion;
Peasant is exchanged for Soldier, which is exchanged for Fugitive, which is
exchanged for Disciple, which is exchanged for Teacher. Champion and
Teacher are not Travellers; they cannot be exchanged for anything. Page and
Peasant can be bought or otherwise gained when being used in a game, but
the other cards cannot, they are not in the Supply. When a non-Supply pile
is empty, that does not count as an empty pile for the game ending
condition or for City (from Prosperity)."

**Comment by soudek01** [comment]
3. Transmute and Vineyard only cost potion, not potion and $0.

4. Again, this isn't important, but I'd say "Possession cont."

5. The Tournament and Prizes tab is one of the most beautiful thing I have
ever seen.

6. Consider treating the Hermit/Madman and Urchin/Mercenary tabs like the
Split piles. Card count would be 10/10 and the text would be Hermit: (what
hermit does) and Madman: (what Madman does).

7. Many of the V2 cards have modified wordings. Here are the links to the
new rulebooks: file:///C:/Users/Sam%20and%20Megan/Downloads/Dominion2nd.pdf
file:///C:/Users/Sam%20and%20Megan/Downloads/Intrigue2nd.pdf

8. The new formatting changes require vanilla bonuses within the text to be
bolded. Example: Nobles, Choose one: *+3 Cards*; or *+2 Actions*.

9. Vanilla bonuses aren't represented in a consistent way. Some of the tabs
have:
"the vanilla bonuses, (card text)."
Others have:
"the vanilla bonuses,
more vanilla bonuses,
then (card text) beneath."
Another example: Look at Bazaar, and then look at Grand Market. Grand
Market is better imo, but they aren't uniform.

10. To make the cards uniform, *all* the vanilla bonuses could be bolded
(including vanilla bonuses within the text) and placed above the card text.
Here is some info from the fan card creation guide on the DS forum:
"If your card offers +Cards, +Actions, +Buys, +$, and/or +VP then list them
in that order. It sounds nitpicky, but all of the official cards adhere to
this strict ordering and have consequently trained our eyes to use
shortcuts to ingest the information. For example, if the first thing on a
card is a boldface "+1 Buy," many experienced players will immediately
understand this to be a terminal action without reading further -- because
the "+1 Action," had it been present, would have appeared above the "+1
Buy" line. Adhering to the standard ordering will guard against players
misreading your cards."
More info at:
http://forum.dominionstrategy.com/index.php?topic=699.0

11. Peddler says +$2.

12. Island says "2 VP" instead of 2 victory symbol.

13. You guys are awesome. I am not trying to be too critical, these are
just suggestions :D!

Sam Oudekerk

**Comment by soudek01** [comment]
14. Sauna/Avanto doesn't have an expansion symbol on the tab.

15. Tabs with multiple cards generate inuniformly (some have: "Card Name:
Card Text" others have: "Card Name:
Card Text Below." Example: "Hovel: When you buy a Victory card, you may
trash this from your hand." vs "Settlers:
+1 Card
+1 Action
 Look through your discard pile. You may reveal a Copper from it and put it
into your hand."
I like Settlers more.

16. Also consider underlining card names (*Settlers:*)

**Comment by wvoigt** [comment]
@soudek01  Thanks again for all the reviews.  Keep them coming.  To help me track things, I'm answering here...

>1 Distant Lands says "4 VP" and "0 VP" instead of 4 {victory symbol} and 0 {victory symbol}.

This is now in the latest branch at View #105

> 2 I am not sure about this, but it seems wierd to me that Page ->Champion and Peasant -> Teacher just say what Page and Peasant do.

I don't think you are looking at the latest branch at View #105  The card text is very similar to what you suggest.

> 3 Transmute and Vineyard only cost potion, not potion and $0.

Still need to work this.  It is not quite as easy as you would think.

> 4 Again, this isn't important, but I'd say "Possession cont."

I just removed it since we no longer have a 2nd Possession card

> 6 Consider treating the Hermit/Madman and Urchin/Mercenary tabs like the Split piles.

Yes, that is the plan.  This will be tracked in Issue #107 

> 7 Many of the V2 cards have modified wordings.

This is now in the latest branch at View #105  For now we only have one card for the two versions.  So I put the new words in brackets [ ] to denote that it is new.  I'm open to other ways of doing this.

> 8 The new formatting changes require vanilla bonuses within the text to be bolded.
> 9 Vanilla bonuses aren't represented in a consistent way.
> 10 To make the cards uniform, *all* the vanilla bonuses could be bolded (including vanilla bonuses within the text) and placed above the card text.

I have opened Issue #113 to track

> 11 Peddler says +$2. 

I do not see this in the latest branch at View #105

> 12 Island says "2 VP" instead of 2 victory symbol.

This is now in the latest branch at View #105

> 14 Sauna/Avanto doesn't have an expansion symbol on the tab. 

I do not see this in the latest branch at View #105

> 15 Tabs with multiple cards generate inuniformly ... (some have: "Card Name: Card Text" others have: 
> 16 Also consider underlining card names (*Settlers:*)

Part of the style guidelines needed.  Tracked in Issue #113

**Comment by wvoigt** [comment]
I updated the code to split out the card types into a json file now.  This completes the refactoring.  All the primary "data" is now in json files.  All the files needed for a full translation are now in the individual language directories.  Along the way, you can now use `--types` to display the card type(s) centered at the top of the body of the divider above the text.

**Comment by wvoigt** [comment]
I'm cross posting this in issue #113 
Here are sample files from the current Card db refactoring pull #105 
Each language (de, en_us, it, nl_du) has 3 files:

- `Test-xx-all.pdf` has a copy of all the dividers
- `Test-xx-groups.pdf` has a copy of the dividers with `--special_card_groups`
- `Test-xx-extras.pdf` has copy of the dividers with `--expansions adventures --expansions empires --expansions promo --expansions base --special_card_groups --exclude_events --exclude_landmarks`

Between the 3 files, I believe every card text option is used for that language.

Please review for any errors you see.  If you see English in a non-English file, it means that there are items that need translation.

[Test-de-all.pdf](https://github.com/sumpfork/dominiontabs/files/617791/Test-de-all.pdf)
[Test-de-extras.pdf](https://github.com/sumpfork/dominiontabs/files/617792/Test-de-extras.pdf)
[Test-de-groups.pdf](https://github.com/sumpfork/dominiontabs/files/617793/Test-de-groups.pdf)
[Test-en_us-all.pdf](https://github.com/sumpfork/dominiontabs/files/617795/Test-en_us-all.pdf)
[Test-en_us-extras.pdf](https://github.com/sumpfork/dominiontabs/files/617796/Test-en_us-extras.pdf)
[Test-en_us-groups.pdf](https://github.com/sumpfork/dominiontabs/files/617801/Test-en_us-groups.pdf)
[Test-it-all.pdf](https://github.com/sumpfork/dominiontabs/files/617800/Test-it-all.pdf)
[Test-it-extras.pdf](https://github.com/sumpfork/dominiontabs/files/617797/Test-it-extras.pdf)
[Test-it-groups.pdf](https://github.com/sumpfork/dominiontabs/files/617798/Test-it-groups.pdf)
[Test-nl_du-all.pdf](https://github.com/sumpfork/dominiontabs/files/617799/Test-nl_du-all.pdf)
[Test-nl_du-extras.pdf](https://github.com/sumpfork/dominiontabs/files/617802/Test-nl_du-extras.pdf)
[Test-nl_du-groups.pdf](https://github.com/sumpfork/dominiontabs/files/617803/Test-nl_du-groups.pdf)

**Comment by sumpfork** [comment]
Deployed.
